### PR TITLE
ip rate limit

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: src/extensions/rate_limit/mod.rs [hunk 1]
```diff
@@ -34,137 +50,81 @@ fn default_jitter_up_to_millis() -> u64 {
...
-            .allow_burst(burst);
-        let limiter = Arc::new(RateLimiter::direct(quota));
-        Self {
-            service,
-            limiter,
-            jitter,
+        // make sure there is at least one rule
+        assert!(!config.rules.is_empty(), "must have at least one rule");
+        // make sure there is at most one ip rule
+        assert!(
```
Comment (xlc) [inline src/extensions/rate_limit/mod.rs:71]: we might as well change the config to something like 
```yml
  rate_limit:
     connection:
       burst: 20 # 20 RPC requests per second per connection
       period_secs: 1
    ip:
       burst: 500 # 500 RPC requests per 10 seconds per ip
       period_secs: 10
```

and then we don't need extra validations for rule count

File: src/extensions/server/mod.rs [hunk 2]
```diff
@@ -90,40 +95,73 @@ impl SubwayServerBuilder {
...
-        Ok((addr, server))
+        let config = self.config.clone();
+
+        let (stop_handle, server_handle) = stop_channel();
+        let handle = stop_handle.clone();
+        let methods = builder().await?;
+
+        // make_service handle each connection
+        let make_service = make_service_fn(move |socket: &AddrStream| {
+            let remote_ip = socket.remote_addr().ip().to_string();
```
Comment (xlc) [inline src/extensions/server/mod.rs:109]: do you know if this respecting to proxy forwarding headers? https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html#x-forwarded-for

File: src/extensions/rate_limit/mod.rs (hunk 1 above)
Comment (ermalkaleci) [reply src/extensions/rate_limit/mod.rs:71]: Yeah makes sense

File: src/extensions/server/mod.rs (hunk 2 above)
Comment (ermalkaleci) [reply src/extensions/server/mod.rs:109]: something I need to explore

**Comment by ermalkaleci** [comment]
@xlc implemented

File: src/extensions/rate_limit/mod.rs [hunk 3]
```diff
@@ -1,17 +1,28 @@
...
+pub use connection::{ConnectionRateLimit, ConnectionRateLimitLayer};
+pub use ip::{IpRateLimit, IpRateLimitLayer};
+pub use xff::XFF;
+
+#[derive(Deserialize, Debug, Clone, Default)]
 pub struct RateLimitConfig {
+    pub ip: Option<Rule>,
+    pub connection: Option<Rule>,
+    #[serde(default)]
+    pub use_xff: Option<bool>,
```
Comment (xlc) [inline src/extensions/rate_limit/mod.rs:21]:
```suggestion
    pub use_xff: bool,
```

File: src/extensions/server/mod.rs [hunk 4]
```diff
@@ -90,40 +96,79 @@ impl SubwayServerBuilder {
...
-        Ok((addr, server))
+        let config = self.config.clone();
+
+        let (stop_handle, server_handle) = stop_channel();
+        let handle = stop_handle.clone();
+        let rpc_module = rpc_module_builder().await?;
+
+        // make_service handle each connection
+        let make_service = make_service_fn(move |socket: &AddrStream| {
+            let remote_ip = socket.remote_addr().ip().to_string();
```
Comment (xlc) [inline src/extensions/server/mod.rs:110]:
```suggestion
            let socket_ip = socket.remote_addr().ip().to_string();
```

to make everything bit more easier to understand

**Pull request merged**